crates/extensions/c8y_auth_proxy/src/server.rs

```diff
@@ -227,28 +369,167 @@ async fn respond_to(
 }
 
 #[cfg(test)]
 mod tests {
     use axum::async_trait;
     use axum::body::Bytes;
+    use axum::headers::authorization::Bearer;
+    use axum::headers::Authorization;
+    use axum::http::Request;
+    use axum::middleware::Next;
+    use axum::TypedHeader;
     use c8y_http_proxy::credentials::JwtRequest;
     use c8y_http_proxy::credentials::JwtResult;
     use c8y_http_proxy::credentials::JwtRetriever;
     use camino::Utf8PathBuf;
     use futures::future::ready;
     use futures::stream::once;
+    use hyper::server::conn::AddrIncoming;
+    use rustls::client::ServerCertVerified;
     use std::borrow::Cow;
     use std::net::Ipv4Addr;
+    use std::net::SocketAddr;
+    use std::time::SystemTime;
     use tedge_actors::Sequential;
     use tedge_actors::Server;
     use tedge_actors::ServerActorBuilder;
     use tedge_actors::ServerConfig;
+    use tokio::net::TcpStream;
+    use tokio_tungstenite::tungstenite::Message;
+    use tokio_tungstenite::Connector;
+    use tokio_tungstenite::MaybeTlsStream;
+    use tokio_tungstenite::WebSocketStream;
 
     use super::*;
 
+    #[tokio::test]
+    async fn forwards_websockets() {
+        let test_app = Router::new().route(
+            "/my/ws/endpoint",
+            get(|ws: WebSocketUpgrade| async { ws.on_upgrade(|_ws| ready(())) }),
+        );
+        let mut server_port = None;
+        for port in 3200..3300 {
+            if let Ok(server) =
+                axum::Server::try_bind(&format!("127.0.0.1:{port}").parse().unwrap())
+            {
+                server_port = Some((server, port));
+            }
+        }
+        let (server, port) = server_port.expect("No port bound");
+        tokio::spawn(server.serve(test_app.into_make_service()));
+        let proxy_port = start_server_port(port, vec!["outdated token", "correct token"]);
+        let (mut ws, _) = connect_to_websocket_port(proxy_port).await;
+        ws.send(Message::Ping("test".as_bytes().to_vec()))
+            .await
+            .expect("Error sending to websocket");
+        assert_eq!(
+            ws.next()
+                .await
+                .unwrap()
+                .expect("Error receiving from websocket"),
+            Message::Pong("test".as_bytes().to_vec())
+        );
+    }
+
+    #[tokio::test]
+    async fn updates_outdated_jwts_for_websocket_connection() {
+        let test_app = Router::new()
+            .route(
+                "/my/ws/endpoint",
+                get(|ws: WebSocketUpgrade| async { ws.on_upgrade(|_ws| ready(())) }),
+            )
+            .layer(axum::middleware::from_fn(auth(|token| {
+                token == "correct token"
+            })));
+        let (server, port) = axum_server();
+        tokio::spawn(server.serve(test_app.into_make_service()));
+        let proxy_port = start_server_port(port, vec!["outdated token", "correct token"]);
+        let (mut ws, _) = connect_to_websocket_port(proxy_port).await;
```

Comment:
Looks like a block of code repeated in the previous test as well, that can be moved to a test setup function that takes the `Router` definition as an argument.